clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryUtils.java

```diff
@@ -175,16 +182,44 @@ public static boolean validateRequiredResourceLabels(Map<String, String> metadat
     }
 
     public static CompressionType preferredCompressionType(List<CompressionType> acceptedCompressionTypes) {
-        // TODO: Support compression in client telemetry.
+        if (acceptedCompressionTypes != null && !acceptedCompressionTypes.isEmpty()) {
+            // Broker is providing the compression types in order of preference. Grab the
+            // first one.
+            return acceptedCompressionTypes.get(0);
+        }
         return CompressionType.NONE;
     }
 
     public static ByteBuffer compress(byte[] raw, CompressionType compressionType) {
-        // TODO: Support compression in client telemetry.
-        if (compressionType == CompressionType.NONE) {
-            return ByteBuffer.wrap(raw);
-        } else {
-            throw new UnsupportedOperationException("Compression is not supported");
+        try {
+            try (ByteBufferOutputStream compressedOut = new ByteBufferOutputStream(512)) {
+                try (OutputStream out = compressionType.wrapForOutput(compressedOut, RecordBatch.CURRENT_MAGIC_VALUE)) {
+                    out.write(raw);
+                    out.flush();
+                }
+                compressedOut.buffer().flip();
+                return ByteBuffer.wrap(Utils.toArray(compressedOut.buffer()));
```

Comment:
Equivalent to below code, the line makes sure only filled bytes are returned from the buffer.

```
ByteBuffer compressedBuffer = ByteBuffer.allocate(bufferOutputStream.buffer().remaining());
compressedBuffer.put(bufferOutputStream.buffer()).flip();
return compressedBuffer;